src/glarea.rs

```diff
@@ -0,0 +1,205 @@
+//! # GLArea Sample
+//!
+//! This sample demonstrates how to use GLAreas and OpenGL
+
+extern crate gtk;
+extern crate libc;
+
+// make moving clones into closures more convenient
+macro_rules! clone {
+    ($($n:ident),+; || $body:block) => (
+        {
+            $( let $n = $n.clone(); )+
+            move || { $body }
+        }
+    );
+    ($($n:ident),+; |$($p:ident),+| $body:block) => (
+        {
+            $( let $n = $n.clone(); )+
+            move |$($p),+| { $body }
+        }
+    );
+}
+
+#[cfg(feature = "opengl")]
+extern crate epoxy;
+
+// Based off of the static-triangle example from gl-rs
+#[cfg(feature = "opengl")]
+mod example {
+    use std::ffi;
+    use std::mem;
+    use std::ptr;
+    use std::cell::RefCell;
+    use std::rc::Rc;
+
+    use gtk;
+    use gtk::traits::*;
+    use gtk::signal::Inhibit;
+    use gtk::{GLArea, Window};
+
+    use epoxy;
+    use epoxy::types::*;
+    use epoxy::Gl;
+
+    fn compile_shader(src: &str, ty: GLenum) -> GLuint {
+        let shader;
+        unsafe {
+            shader = Gl.CreateShader(ty);
+            // Attempt to compile the shader
+            let csrc = ffi::CString::new(src).unwrap().as_ptr();
+            Gl.ShaderSource(shader, 1, &csrc, ptr::null());
+            Gl.CompileShader(shader);
+
+            // Get the compile status
+            let mut status = epoxy::FALSE as GLint;
+            Gl.GetShaderiv(shader, epoxy::COMPILE_STATUS, &mut status);
+
+            // Fail on error
+            if status != (epoxy::TRUE as GLint) {
+                let mut len = 0;
+                Gl.GetShaderiv(shader, epoxy::INFO_LOG_LENGTH, &mut len);
+                let mut buf = vec![0u8; len as usize - 1];
+                Gl.GetShaderInfoLog(shader, len, ptr::null_mut(), buf.as_mut_ptr() as *mut GLchar);
+                panic!("Error compiling shader");
+            }
+        }
+        shader
+    }
+
+    fn link_program(vs: GLuint, fs: GLuint) -> GLuint { unsafe {
+        let program = Gl.CreateProgram();
+        Gl.AttachShader(program, vs);
+        Gl.AttachShader(program, fs);
+        Gl.LinkProgram(program);
+
+        // Get the link status
+        let mut status = epoxy::FALSE as GLint;
+        Gl.GetProgramiv(program, epoxy::LINK_STATUS, &mut status);
+
+        // Fail on error
+        if status != (epoxy::TRUE as GLint) {
+            let mut len: GLint = 0;
+            Gl.GetProgramiv(program, epoxy::INFO_LOG_LENGTH, &mut len);
+            let mut buf = vec![0u8; len as usize - 1];
+            Gl.GetProgramInfoLog(program, len, ptr::null_mut(), buf.as_mut_ptr() as *mut GLchar);
+            panic!("Error linking shader");
+        }
+
+        program
+    }}
+
+    pub fn main() {
+        if gtk::init().is_err() {
+            println!("Failed to initialize GTK.");
+            return;
+        }
+
+        let window = Window::new(gtk::WindowType::Toplevel).unwrap();
+
+        let cell = Rc::new(RefCell::new(GLArea::new().unwrap()));
```

Comment:
Any `GObject` already is a refcounted shared reference so you don't need to add another layer of that. Cloning will just add to refcount and all methods take `&self`.